### PR TITLE
Fixes #51: Synchronization strategy selector.

### LINKED ISSUE #51: Add a selector for conflicts resolution strategy

### PR DESCRIPTION
**Pull request by n1k0**
WiP, refs #51.

This introduces a new dropdown menu for the synchronization strategy to use next to the *Synchronize* button:

 

Feedback=? @leplatrem @glasserc @Natim

### CODE REVIEW

**Comment by Natim** [comment]
\o/ really nice looking

**Comment by glasserc** [comment]
That's clever, but at first glance I expected it to set the strategy, not just sync once with this strategy.

**Comment by Natim** [comment]
> That's clever, but at first glance I expected it to set the strategy, not just sync once with this strategy.

We could use the config to set the default strategy but still let the user choose a different one for some edge cases.

**Comment by leplatrem** [comment]
nit: `Manual` could be `Manual resolution` ?

**Comment by n1k0** [comment]
> That's clever, but at first glance I expected it to set the strategy, not just sync once with this strategy.

I can probably persist the selected strategy in local state and make the button group a kind of toggler. I'll toy around with the idea, that's useful thanks!

**Comment by n1k0** [comment]
> nit: `Manual` could be `Manual resolution` ?

Yeah, wording is poor here, but these concepts carry a bunch of concepts so it's hard to reflect them with simple labels. *Manual resolution* will do for now, thanks.

**Comment by n1k0** [comment]
> We could use the config to set the default strategy but still let the user choose a different one for some edge cases.

Not sure a default strategy is really applicable to a given collection, but that's an option for sure. I'll probably start with persisting current selected strategy first.

**Comment by n1k0** [comment]
Updated the patch so selected strategy is persisted in local comp state, much better imho:

**Comment by n1k0** [comment]
Gah, I need to move state handling to the upper level as the two buttons don't show the same selected strategy now :D

**Comment by n1k0** [comment]
>  I need to move state handling to the upper level as the two sync buttons don't show the same selected strategy now

Done:

**Comment by n1k0** [comment]
Wrote simple tests for this, can now probably be reviewed. r=? @glasserc @Natim @leplatrem 

Note: might sound obvious but please ensure testing the UI running the patch locally while reviewing.

**Comment by n1k0** [comment]
You didn't ask but I improved the look of the button:



I really like how it's now very explicit what is the current sync strategy.

**Comment by n1k0** [comment]
Added tooltips:



Please review wording.

File: scripts/components/CollectionList.js [hunk 1]
```diff
@@ -162,26 +167,101 @@ class Table extends Component {
...
+    this.state = {open: false};
+  }
+
+  static strategies = {
+    [MANUAL]: {
+      label: "MANUAL resolution",
+      help: "Do not automatically resolve conflicts.",
+    },
+    [CLIENT_WINS]: {
+      label: "CLIENT_WINS resolution",
```
Comment (leplatrem) [inline scripts/components/CollectionList.js]: Why not «Client wins» ?

File: scripts/components/CollectionList.js [hunk 2]
```diff
@@ -162,26 +167,101 @@ class Table extends Component {
...
+class SyncButton extends Component {
+  constructor(props) {
+    super(props);
+    this.state = {open: false};
+  }
+
+  static strategies = {
+    [MANUAL]: {
+      label: "MANUAL resolution",
+      help: "Do not automatically resolve conflicts.",
```
Comment (leplatrem) [inline scripts/components/CollectionList.js]: «Prompt when a conflict occurs.» ?

File: scripts/components/CollectionList.js [hunk 3]
```diff
@@ -162,26 +167,101 @@ class Table extends Component {
...
+      help: "Overwrite local conflicting records with remote data.",
+    },
+  }
+
+  openMenu = () => this.setState({open: true});
+  closeMenu = () => this.setState({open: false});
+
+  doSync = () => {
+    this.props.sync();
+    this.setState({open: false});
```
Comment (leplatrem) [inline scripts/components/CollectionList.js]: `this.closeMenu()` ?

File: scripts/components/CollectionList.js [hunk 4]
```diff
@@ -162,26 +167,101 @@ class Table extends Component {
...
+  };
+
+  selectStrategy = (strategy) => {
+    return (event) => {
+      event.preventDefault();
+      this.props.selectStrategy(strategy);
+      this.setState({open: false});
```
Comment (leplatrem) [inline scripts/components/CollectionList.js]: nitto?

Comment (n1k0) [inline scripts/components/CollectionList.js]: > nitto?

Tristan?

File: scripts/components/CollectionList.js [hunk 5]
```diff
@@ -162,26 +167,101 @@ class Table extends Component {
...
+
+  render() {
+    const {strategy} = this.props;
+    const {open} = this.state;
+    return (
+      <div className={`sync-btn-group btn-group ${open ? "open" : ""}`}>
+        <button type="button" className="btn btn-info btn-sync"
+          onClick={this.doSync}>
+          <span className="caption">Synchronize</span>
+          <span className="label label-primary">{strategy.toUpperCase()}</span>
```
Comment (leplatrem) [inline scripts/components/CollectionList.js]: `strategy.toUpperCase().replace("_", "")` ?

File: scripts/components/CollectionList.js [hunk 6]
```diff
@@ -162,26 +167,101 @@ class Table extends Component {
...
+        </button>
+        <button type="button" className="btn btn-info dropdown-toggle"
+          title="Select alternative strategy"
```
Comment (leplatrem) [inline scripts/components/CollectionList.js]: «Select conflict resolution strategy» ?

File: scripts/components/CollectionList.js [hunk 7]
```diff
@@ -162,26 +167,101 @@ class Table extends Component {
...
-function ListActions(props) {
-  const { name, onSyncClick, onResetSyncClick } = props;
+class SyncButton extends Component {
+  constructor(props) {
+    super(props);
+    this.state = {open: false};
+  }
+
+  static strategies = {
+    [MANUAL]: {
```
Comment (leplatrem) [inline scripts/components/CollectionList.js]: Maybe add a `buttonLabel` key here instead of uppercasing in the widget render

**Comment by leplatrem** [comment]
It looks good to me! Most of my comments are ideas/feedback that can be ignored if irrelevant.

File: scripts/components/CollectionList.js (hunk 4 above)
Comment (leplatrem) [inline scripts/components/CollectionList.js]: >> nitto?
>
> Tristan?

smart laser-guided cleaning ?

File: scripts/components/CollectionList.js (hunk 7 above)
Comment (n1k0) [inline scripts/components/CollectionList.js]: I'm confused, there's a `label` property just below.

File: scripts/components/CollectionList.js (hunk 1 above)
Comment (n1k0) [inline scripts/components/CollectionList.js]: I went with reusing the Kinto.js documented term, but yeah, why not, I don't know.

**Comment by n1k0** [comment]
Now I wish I could have a r+ somehow :/

**Comment by Natim** [comment]
r+

**Pull request merged**